### PR TITLE
Added parallel searching

### PR DESCRIPTION
**Pull request by eshvatskyi**

### CODE REVIEW

**Comment by richardschneider** [comment]
Can you please fix the warnings and errors, see the [build log]().

File: src/WireReader.cs [hunk 1]
```diff
@@ -201,11 +201,11 @@ public string ReadDomainName()
 
             // Read current label and remaining labels.
             var buffer = ReadBytes(length);
-            if (buffer.Any(c => c > 0x7F))
-            {
-                throw new InvalidDataException("Only ASCII characters are allowed.");
-            }
-            var name = Encoding.ASCII.GetString(buffer, 0, length);
+            //if (buffer.Any(c => c > 0x7F))
```
Comment (richardschneider) [inline src/WireReader.cs]: Why are you removing this check. The spec only allows US-ASCII.

File: src/WireReader.cs [hunk 2]
```diff
@@ -293,7 +293,7 @@ public TimeSpan ReadTimeSpan32()
         public IPAddress ReadIPAddress(int length = 4)
         {
             var address = ReadBytes(length);
-            return new IPAddress(address);
+            return new IPAddress(address.ToArray());
```
Comment (richardschneider) [inline src/WireReader.cs]: Why the change? `address` is already a byte array.

File: src/WireWriter.cs [hunk 3]
```diff
@@ -272,10 +272,10 @@ public void WriteDomainName(string name, bool uncompressed = false)
...
         public void WriteString(string value)
         {
-            if (value.Any(c => c > 0x7F))
-            {
-                throw new ArgumentException("Only ASCII characters are allowed.");
-            }
+            //if (value.Any(c => c > 0x7F))
+            //{
+            //    throw new ArgumentException("Only ASCII characters are allowed.");
+            //}
```
Comment (richardschneider) [inline src/WireWriter.cs]: Why are you removing this check. The spec only allows US-ASCII.

File: src/WireReader.cs (hunk 1 above)
Comment (eshvatskyi) [reply src/WireReader.cs]: yes, when you Encoding.ASCII.GetString - it replaces all non ASCII with "?", so looks legal to just send no answer. But I left it for writing because then it illegal :)

Comment (eshvatskyi) [reply src/WireReader.cs]: Any way https://tools.ietf.org/html/rfc7997

File: src/WireReader.cs (hunk 2 above)
Comment (eshvatskyi) [reply src/WireReader.cs]: Sorry will remove, played with Span<byte>, but this gives not much for using in this project.
see System.Memory

File: src/WireWriter.cs (hunk 3 above)
Comment (eshvatskyi) [reply src/WireWriter.cs]: same as prev

File: src/WireReader.cs (hunk 1 above)
Comment (richardschneider) [reply src/WireReader.cs]: RFC 7997 does not apply to DNS.  DNS strings and labels have very precise requirements.

Please restore the checks.

File: src/WireReader.cs (hunk 2 above)
Comment (richardschneider) [reply src/WireReader.cs]: Yes, `Span` could be helpful.  However, it is not natively available for all the targets we support.

**Comment by richardschneider** [comment]
Parallel searching seems to be a case of [premature optimization](https://softwareengineering.stackexchange.com/a/80092/80038). Do you have any metrics?

Most DNS requests have just 1 question.  The overhead of using `Parallel.ForEach` will far exceed any benefits gained in processing multiple questions.

File: src/Resolving/NameServer.cs [hunk 4]
```diff
@@ -36,18 +35,11 @@ public class NameServer : IResolver
...
-            CancellationToken cancel = default(CancellationToken))
+        public Task<Message> ResolveAsync(Message request, CancellationToken cancel = default(CancellationToken))
         {
             var response = request.CreateResponse();
 
-            foreach (var question in request.Questions)
-            {
-                await ResolveAsync(question, response, cancel);
-                if (response.Answers.Count > 0 && !AnswerAllQuestions)
-                    break;
```
Comment (richardschneider) [inline src/Resolving/NameServer.cs]: This test is really required.  The DNS spec states that when multiple questions are present only one question is answered.  The mDNS spec states that all questions are answerex.

File: src/Resolving/NameServer.cs [hunk 5]
```diff
@@ -96,20 +85,25 @@ public class NameServer : IResolver
         ///   If the question's domain does not exist, then the closest authority
         ///   (<see cref="SOARecord"/>) is added to the <see cref="Message.AuthorityRecords"/>.
         /// </remarks>
-        public async Task<Message> ResolveAsync(Question question, Message response = null, CancellationToken cancel = default(CancellationToken))
+        public Message Resolve(Question question, Message response = null)
         {
```
Comment (richardschneider) [inline src/Resolving/NameServer.cs]: This is a **breaking change**.  There are a number of projects already using this package.

File: src/Resolving/NameServer.cs (hunk 4 above)
Comment (eshvatskyi) [reply src/Resolving/NameServer.cs]: first question or first found? Im asking because it's very important )

Comment (richardschneider) [reply src/Resolving/NameServer.cs]: I think DNS is dark on this.  Most DNS callers will never send a multi-question request, see https://serverfault.com/questions/742785/multi-query-multiple-dns-record-types-at-once

However, mDNS callers use it to get the all the `A/AAAA` records at the same time.

File: src/Resolving/NameServer.cs (hunk 5 above)
Comment (eshvatskyi) [reply src/Resolving/NameServer.cs]: Returned method for back compatibility,
one notice IResolver does not provide such method, should it?

File: src/Resolving/NameServer.cs (hunk 4 above)
Comment (eshvatskyi) [reply src/Resolving/NameServer.cs]: Added back compatibility, if AnswerAllQuestions then Parallel otherwise as it was

**Comment by richardschneider** [comment]
The appveyor build failed because `chocolatey` was down for 17 hours.  I've resubmitted.

**Comment by richardschneider** [comment]
Closing PR without merge because of lack of interest
